firmware/hackrf_usb/sgpio_m0.s

```diff
@@ -1,108 +1,673 @@
+	length .req r0
+	mov length, shortfall_length                    // length = shortfall_length            // 1
+	cmp length, #0                                  // if length == 0:                      // 1
+	beq idle                                        //      goto idle                       // 3
+
+	// Otherwise, roll back the state to ignore the current shortfall, then jump to idle.
+	prev .req r0
+	ldr prev, [state, #PREV_LONGEST_SHORTFALL]      // prev = prev_longest_shortfall        // 2
+	str prev, [state, #LONGEST_SHORTFALL]           // state.longest_shortfall = prev       // 2
+	ldr prev, [state, #NUM_SHORTFALLS]              // prev = num_shortfalls                // 2
+	sub prev, #1                                    // prev -= 1                            // 1
+	str prev, [state, #NUM_SHORTFALLS]              // state.num_shortfalls = prev          // 2
+
+	b idle                                          // goto idle                            // 3
+
+tx_start:
+
+	// Reset counts.
+	reset_counts                                    // reset_counts()                       // 10
+
+tx_loop:
+
+	// Wait for and clear SGPIO interrupt.
+	await_sgpio tx                                  // await_sgpio()                        // 34
+
+	// Check if a return to idle mode was requested.
+	// If so, we may need to roll back shortfall stats.
+	mode .req r3
+	ldr mode, [state, #MODE]                        // mode = state.mode                    // 2
+	cmp mode, #MODE_IDLE                            // if mode == IDLE:                     // 1
+	beq checked_rollback                            //      goto checked_rollback           // 1 thru, 3 taken
+
+	// Check if there is enough data in the buffer.
+	//
+	// The number of bytes in the buffer is given by (m4_count - m0_count).
+	// We need 32 bytes available to proceed. So our margin, which we want
+	// to be positive or zero, is:
+	//
+	// buf_margin = m4_count - m0_count - 32
+	//
+	// If there is insufficient data, transmit zeros instead.
+	buf_margin .req r0
+	ldr buf_margin, [state, #M4_COUNT]              // buf_margin = m4_count                // 2
+	sub buf_margin, count                           // buf_margin -= count                  // 1
+	sub buf_margin, #32                             // buf_margin -= 32                     // 1
+	bmi tx_zeros                                    // if buf_margin < 0: goto tx_zeros     // 1 thru, 3 taken
+
+	// Update buffer pointer.
+	update_buf_ptr                                  // update_buf_ptr()                     // 3
+
+	// At this point we know there is TX data available.
+	// If still in TX start mode, switch to TX run.
+	cmp mode, #MODE_TX_START                        // if mode != TX_START:                 // 1
+	bne tx_write                                    //      goto tx_write                   // 1 thru, 3 taken
+	mov mode, #MODE_TX_RUN                          // mode = TX_RUN                        // 1
```

Comment:
This could clobber a new mode that was just set by the M4.

Reply:
Fixed in cb3cfe8